Introduce bundles content type in the repository

RohitTopi opened the pull request:
Introduce bundles content type

To improve cold cache loading, several files which are known to be required in sets can be bundled together and streamed as an object pack. This is the case, for instance, for well-known Python libraries (e.g. Tensorflow) that upon start always load a certain set of inter-dependent python files.

To support this, a new content type *bundles* is introduced. A bundle can have the binary format of object packs used by the gateway as a form to concatenate multiple objects in a single BLOB. Such bundles would be created using a special hidden file (.cvmfsbundles in [draft PR 2736](https://github.com/cvmfs/cvmfs/pull/2736)). Bundle files can be chunked the same way regular large files get chunked. Chunked and non-chunked bundle files would get a B as a hash suffix.

Corresponding JIRA issue: [CVM-1886](https://sft.its.cern.ch/jira/browse/CVM-1886)

RohitTopi committed:
Create bundle.cc

jblomer commented on cvmfs/bundle.cc:
```diff
@@ -0,0 +1 @@
+
```
Let's add least add the usual cvmfs comment header.

jblomer reviewed (changes requested):
Let's add to this one the header file and a skeleton for the unit test. Let's also add the files to the CMakeLists.txt file.

RohitTopi committed:
Create bundle.h

RohitTopi committed:
Added CVM-FS comment header

RohitTopi committed:
Add bundle.cc

RohitTopi committed:
Create t_bundle.cc

RohitTopi committed:
Add t_bundle.cc

RohitTopi committed:
Add Bundle unit test stub

RohitTopi committed:
Add Bundle class

RohitTopi committed:
Add CreateBundle() function

jblomer commented on cvmfs/bundle.cc:
```diff
@@ -0,0 +1,87 @@
+#include "util/posix.h"
+#include "util/string.h"
+
+ObjectPack* Bundle::CreateBundle(const JSON* json_obj)
```
Style: 

```suggestion
ObjectPack *Bundle::CreateBundle(const JSON *json_obj) {
```

jblomer commented on cvmfs/bundle.h:
```diff
@@ -0,0 +1,39 @@
+#include "pack.h"
+
+/**
+ * A bundle has the binary format of object packs used by the gateway as
```
```suggestion
 * A bundle has the binary format of object packs used as
```

jblomer commented on cvmfs/bundle.h:
```diff
@@ -0,0 +1,39 @@
+ * if the file size of any of the files exceeds a maximum value kMaxFileSize
+ * The JSON Object json_obj passed has a bundle id and a JSON array of file
+ * paths
+ *
```
Let's document the JSON structure that `CreateBundle` expects

jblomer commented on cvmfs/bundle.h:
```diff
@@ -0,0 +1,39 @@
+ * Returns a pointer to the created ObjectPack on success, NULL
+ * otherwise
+ */
+  ObjectPack* CreateBundle(const JSON* json_obj);
```
```suggestion
  ObjectPack *CreateBundle(const JSON *json_obj);
```

jblomer commented on cvmfs/bundle.cc:
```diff
@@ -0,0 +1,87 @@
+    PANIC(kLogStderr, "JSON object not found");
+  } else {
+    // create an ObjectPack
+    ObjectPack* op = new ObjectPack();
```
Let's use a `UniquePtr` from pointer.cc. We then don't need the explicit `delete` calls in the error paths. We return it with `UniquePtr::Release()`

jblomer commented on cvmfs/bundle.cc:
```diff
@@ -0,0 +1,87 @@
+
+      // if file doesn't exist then bundle is not created
+      if (!FileExists(filepath)) {
+        PrintWarning("File not found: " + filepath);
```
We should decide here: either warn and skip (instead of returning) or we `PANIC`. I think we want to warn and skip.

jblomer commented on cvmfs/bundle.cc:
```diff
@@ -0,0 +1,87 @@
+
+    value = (value->first_child);  // bundleid
+    value = (value->next_sibling);  // JSON array
+    if (value->first_child) {
```
Combine with while loop further down to remove the code duplication.

jblomer commented on cvmfs/bundle.cc:
```diff
@@ -0,0 +1,87 @@
+          return NULL;
+        }
+      }
+    } else {
```
Let's error out early if an empty bundle is found.

jblomer reviewed (changes requested):
It's taking shape!  You can actually add unit tests for `CreateBundle` right away.

RohitTopi committed:
Style fixes

RohitTopi committed:
Update class comment

RohitTopi committed:
Update function comment

Document the JSON structure expected by the CreateBundle() function

RohitTopi committed:
Error out early if bundle would be empty

RohitTopi committed:
Refactor code to avoid code duplication

RohitTopi committed:
Allow creating bundle with missing files

If one or more files are not found then those files are skipped with a
warning

RohitTopi committed:
Remove explicit delete calls from the error paths

RohitTopi committed:
Add function to parse bundle specification

RohitTopi committed:
Add test for parsing bundle specification

RohitTopi committed:
Update CreateBundle() function

Removed redundant JSON Parsing which, is done by the new function
ParseBundleSpec(). The function CreateBundle() takes in a set of
strings as input instead of a JSON object.
Additionally, replaced the function call AddFileToObjectPack() with
actual code.

RohitTopi committed:
Add test for creation of a bundle

RohitTopi committed:
Style fixes

Fixed indentation

RohitTopi committed:
Add a separate function for parsing

Added a separate function ParseBundleSpecFile() to parse the bundle
specification file

RohitTopi committed:
Move ParseBundleSpecFile() to Bundle class

Make the ParseBundleSpecFile() a static member function of the Bundle
class

RohitTopi committed:
Don't add large files to the bundle

Skip the files that have sizes exceeding the kMaxFileSize

RohitTopi committed:
Avoid creating empty bundles

RohitTopi committed:
Style fix

RohitTopi committed:
Merge branch 'cvmfs:devel' into CVM-1886-bundle_file

RohitTopi committed:
Minor fixes

Fix type conversion errors

RohitTopi committed:
Pass the object by reference

RohitTopi committed:
Create bundle objects using the Bundle class

Remove code that was used to create JSON document.
The content of the .cvmfsbundles file is now parsed by the
ParseBundleSpecFile() function and then bundle objects are created
using CreateBundle() function.

RohitTopi committed:
Add PublishBundlesCallback() function

RohitTopi committed:
Upload bundles to storage

RohitTopi committed:
Add BundleEntry struct

RohitTopi committed:
Add bundles to the catalog

RohitTopi committed:
Add support for inserting bundle entry

Add SqlBundleInsert class that will allow inserting bundle entries
into the bundles table

RohitTopi committed:
Merge branch 'cvmfs:devel' into CVM-1886-bundle_file

RohitTopi committed:
Use FilepathSet alias for set of filepaths

RohitTopi committed:
Update bundle id for each file in a bundle

RohitTopi committed:
Add support for update of bundle id

RohitTopi committed:
Use std namespace

RohitTopi committed:
Initialize the bundle name of a bundle entry

RohitTopi committed:
Minor fixes

RohitTopi committed:
Style fixes

vvolkl commented:
Superseded by #4002
